### DISCUSSION TITLE
OG tags are not working with dynamic content

### DISCUSSION THREAD

**Discussion by M1ck0** [vercel/next.js#12350] [Help] [answered]
OG tags used by Facebook and Twitter when sharing page are not working

## To Reproduce

I tried two things. Using ```next-seo``` library to set OG tags. Thing is that dynamic content in OG tags is not displayed at all using ```next-seo```. Currently I don't have sample code for ```next-seo``` but I used same one that is displayed in [DOCS](https://www.npmjs.com/package/next-seo). 

Next I tried with ```Head``` from ```next/head``` but it doesn't seem to work

```jsx
<Head>
  <title>{review.title}</title>
  <meta name="description" content={review.review.pageDescription}/>
  <meta name="og:url" content={`https://website.com${router.asPath}`}/>
  <meta name="og:title" content={review.title}/>
  <meta name="og:description" content={review.content.slice(0, 50)}/>
  <meta name="og:image" content={review.featuredImage.sourceUrl}/>
  <meta name="og:site_name" content='site_name'/>
</Head>
```

## Expected behavior

```<meta>``` should look like this 
 ```html
<meta property="og:image" content="https://website.com/images/image.jpg/>
```
but instead it looks like this 
```html
<meta property="og:image"/>
```

## System information

- OS: Windows
- Version of Next.js: 9.3
- Version of Node.js: 12.16.1

**Comment by timneutkens** [comment]
How are you fetching data? Maybe you're fetching data client-side?

**Reply by M1ck0** [reply]
I am using Apollo and GraphQL. Data is present in View Page Source as JSON object in
 ```html
<script id="__NEXT_DATA__" type="application/json">
```

**Reply by priyanshu-mq** [reply]
> How are you fetching data? Maybe you're fetching data client-side?

I am facing the same issue, and yes i am fetching data from client side, although tags are reflecting in inspector with data but not in page source or og checker

![Screenshot_1](https://user-images.githubusercontent.com/85607051/131687379-589c6361-3948-4b73-8de2-34af6bc74ebe.png)

**Comment by M1ck0** [comment]
Anyone? I am still having this issue

**Reply by c0b41** [reply]
hard to say something, show some code maybe.

**Comment by M1ck0** [comment]
I am using ```withApollo``` HOC to export pages that use GraphQL

This is how I export page
```
export default withApollo({ ssr: true })(Reviews)
```

This is how I call query to fetch data

```
const { loading, data } = useQuery(single_review_query, {
    variables: { slug: slug },
    ssr: true
  })
```

Data is fetched and displayed in View page source of website as JSON object between ```<script id="__NEXT_DATA__" type="application/json"></script>``` tags. Every part of data is extracted as it should except from ```<meta>``` tags.

Do you need any more code or explanation?

**Reply by brianlovin** [reply]
It's really hard to say without seeing a bit more code. My hunch based on what you've provided here is that there's something going on with the way data is being queried on the client that might not be updating the values. Have you dug into the debugger or console logging the values before they hit the `<NextSEO>` or `<Head>` components to see what they are (probably `undefined`?)

**Reply by M1ck0** [reply]
Yes. It is undefined on first render. And in View Page Source those tags are empty.

**Reply by brianlovin** [reply]
Can you start working your way up the data flow and figure out where the data isn't being set?

**Reply by M1ck0** [reply]
Here is what I think. When I call query it is called as it should be and when I ```console.log(data)``` it returns data. not undefined.

Then I have ```useEffect``` that triggers when ```data``` from query changes and in that useEffect I setState so I can use that state through app to populate it with content. 

Could it be that useEffect is a problem in this case?

**Reply by brianlovin** [reply]
It might be useful if you could spin up a small demo next.js app using your setup, but without any sensitive bits, so we can try to repro on our end and debug together.

**Reply by M1ck0** [reply]
I fixed it.. I posted an answer. But I can setup repo with broken and working code if someone might find it useful

**Comment by M1ck0** [comment, marked as answer]
```useEffect``` is problem in this case.. If I use ```data.review.title```, instad of setting ```data``` to ```state``` and then using it from there, it works. How stupid am I.. damn.. Thank you guys :)

**Reply by bymoe** [reply]
@M1ck0 I'm having the exact same issue, Can I please see this bit - `useEffect` - of your code? Thank you.

**Reply by Badrouu17** [reply]
> `useEffect` is problem in this case.. If I use `data.review.title`, instad of setting `data` to `state` and then using it from there, it works. How stupid am I.. damn.. Thank you guys :)

 @M1ck0 Can you share with us the solution, please!

**Reply by M1ck0** [reply]
Yes sure. Give me few minutes.

**Reply by M1ck0** [reply]
First of all sorry for late answer.

```jsx
const Home = (data) => {
  return (
    /* Use data directly in page instead of setting it to state and then using it from state*/
  )
}

export async function getServerSideProps() {
  return {
    props: {
      data: /* data that you fetched */
    }
  }
}
```

**Comment by drasya** [comment]
I've facing same issue, Please help me over here.

**Reply by bymoe** [reply]
I fixed mine by using the meta tags right from the props not by storing them in a state first.

**Reply by M1ck0** [reply]
Read the accepted answer

**Reply by drasya** [reply]
Thank you so much!

**Reply by cazterk** [reply]
can i see the code?

**Comment by dmbrooking** [comment]
So I'm doing SSR, accessing the props directly, instead of from state - what the accepted answer is doing.. and it's still not reading my tags.

My code is:

```
export const getServerSideProps: GetServerSideProps = async ({ params: { slug } }) => {
  .... var setup ....
  const [postData, topPostsData] = await Promise.all([
    wordpressClient.query({
      query: GET_POST,
      variables: {
        id: uri
      }
    }),
    wordpressClient.query({
      query: GET_TOP_POSTS,
      variables: {
        year: twoWeeksAgo.getFullYear(),
        month: twoWeeksAgo.getMonth() + 1,
        day: twoWeeksAgo.getDate()
      }
    })
  ]);

  const { post } = postData.data;

  return {
    props: {
      post
    }
  };
};
```

and then in my page: 
```
const PostPage: NextPage<Props> = ({ post }) => {
  const { publicRuntimeConfig } = getConfig();

  if (!post) return <DefaultErrorPage statusCode={404} />;

  return (
    <PostPageStyled>
      <Helmet>
        <meta property="og:title" content={post.title} />
        <meta property="og:url" content={`${publicRuntimeConfig.BASE_URL}${post.uri}`} />
        <meta property="og:description" content={post.excerpt ? post.excerpt : post.title} />
        <meta property="og:image" content={post.featuredImage.node.sourceUrl} />
        <meta property="og:image:secure_url" content={post.featuredImage.node.sourceUrl} />
        <meta property="og:image:width" content={`${post.featuredImage.node.mediaDetails.width}`} />
        <meta property="og:image:height" content={`${post.featuredImage.node.mediaDetails.height}`} />
        <meta name="twitter:title" content={post.title} />
        <meta name="twitter:description" content={post.excerpt ? post.excerpt : post.title} />
        <meta name="twitter:image" content={post.featuredImage.node.sourceUrl} />
      </Helmet>

      ....
```
but it's not seeing any of that stuff.  Am I missing something?

**Reply by MonstraG** [reply]
Yes, `publicRuntimeConfig` is only filled correctly when using `getInitialProps`: https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration
```
A page that relies on publicRuntimeConfig must use getInitialProps to opt-out of Automatic Static Optimization. Runtime configuration won't be available to any page (or component in a page) without getInitialProps.
```

**Comment by SirReaDy** [comment]
I am facing same issue on dynamic pages.
I am fetching my data directly in the SSR

The meta tags are visible in the page source code and are correct. but are not picked by og check or metatags.io

```
export const getServerSideProps: GetServerSideProps = async context => {
const {data: userData} = await graphqlRequestHandler(
    queries.userData,
    {
      username: context.query.username,
    },
  );
  const user = userData.data.userData
 return {
    props: {
      user ,
    },
  };
}
```


and inside my component I use
```
const UserPage = ({user}) =>{
...
...
   <meta name='title' content={user.name} key='title' />
    <meta name='description' content={user.description} key='description' />
...
...
```
I do not use publicRuntimeConfig 
Using the same setup in a non-dynamic/slug page works as intended.

**Reply by cazterk** [reply]
`useEffect(() => {
    setIsMounted(true);
  }, [])`

I had to remove that line in my _app.tsx and things worked out

**Reply by SirReaDy** [reply]
I do not have any loading states in my _app file. The only thing is that I display my layout conditionally with **getlayout** based on the page., and I have my contexts there.

**Reply by cazterk** [reply]
Alright probably a different cause for you

**Reply by SirReaDy** [reply]
It's really weird as nothing shows me it should not work, again, the same setup on a non-slug page works perfectly, same code. I'm not sure if its had something to do that I am using graphql for data fetching and not rest.

**Reply by cazterk** [reply]
Are they showing up in the browser dev tools? `<head></head>`

**Reply by SirReaDy** [reply]
So they are in the head dev tools, but not in the view page source.
They are present in both places for the non-dynamic page

**Reply by cazterk** [reply]
That's exactly what I was facing

**Reply by M1ck0** [reply]
You shouldn't use data directly from `({user})`, but rather set that data into the state, and use it from the state.

```
const UserPage = ({user}) =>{
  const [data, setData] = useState({})

  useEffect(() => {
    setData(user)
  }, [user])
...
<meta name='title' content={data.name} key='title' />
```

Have you tried this? If not, try it and see if it is working.

If that's not the case, maybe you need to use `next-seo` package for that. I think it might be a problem.

**Reply by SirReaDy** [reply]
Thanks for your answer. Unfortunally none worked. Tried both using the state and the package, no results.

**Reply by SirReaDy** [reply]
Update, the issue was a loading state coming from a session check. Thanks everyone

**Reply by cazterk** [reply]
Nice to hear 🤜🏼🤛🏽

**Reply by script8888** [reply]
> Update, the issue was a loading state coming from a session check. Thanks everyone

what was your solution to that? I believe that might be same issue with me
